qualtran/bloqs/controlled_state_preparation.py

```diff
@@ -0,0 +1,469 @@
+            rot_reg=rot_reg,
+            phase_gradient=phase_gradient,
+        )
+        bb.add(OneEffect(), q=rot_ancilla)
+        return control, target_state, rot_reg, phase_gradient
+
+    def _get_phase_ROM_values(self, amplitude_rom_vals):
+        """As we are using the equivalent to controlled Z to do the rotations instead of Rz, there
+        is a phase offset for each coefficient that has to be corrected. This offset is half of the
+        turn angle applied, and is added to the phase for each coefficient.
+        """
+        offset_angles = [0] * (2**self.n_qubits)
+        for i in range(self.n_qubits):
+            for j in range(2**i):
+                item_range = 2 ** (self.n_qubits - i)
+                # if the rom has value 0 the formula gives 180, when it should be 0
+                if amplitude_rom_vals[i][j] == 0:
+                    offset = 0
+                else:
+                    offset = np.pi * amplitude_rom_vals[i][j] / (2**self.rot_reg_size)
+                for k in range(item_range * j, item_range * (j + 1)):
+                    offset_angles[k] += offset
+        # if the matrix is the adjoint, the angles have to be undone, thus just load -theta
+        if self.uncompute:
+            angles = [offset - np.angle(c) for c, offset in zip(self.state, offset_angles)]
+        else:
+            angles = [np.angle(c) - offset for c, offset in zip(self.state, offset_angles)]
+        rom_values = [RotationTree.angle_2_ROM_value(a, self.rot_reg_size) for a in angles]
+        return rom_values
+
+
+@bloq_example
+def _controlled_state_preparation():
+    state_coefs = (
+        (-0.42677669529663675 - 0.1767766952966366j),
+        (0.17677669529663664 - 0.4267766952966367j),
+        (0.17677669529663675 - 0.1767766952966368j),
+        (0.07322330470336305 - 0.07322330470336309j),
+        (0.4267766952966366 - 0.17677669529663692j),
+        (0.42677669529663664 + 0.17677669529663675j),
+        (0.0732233047033631 + 0.17677669529663678j),
+        (-0.07322330470336308 - 0.17677669529663678j),
+    )
+    return ControlledStatePreparationUsingRotations(n_qubits=2, rot_reg_size=4, state=state_coefs)
+
+
+_CONTROLLED_STATE_PREP_DOC = BloqDocSpec(
+    bloq_cls=ControlledStatePreparationUsingRotations,
+    import_line='from qualtran.bloqs.controlled_state_preparation import ControlledStatePreparationUsingRotations',
+    examples=(_controlled_state_preparation,),
+)
+
+
+@attrs.frozen
+class ControlledQROMRotateQubit(Bloq):
+    r"""Array of controlled rotations $Z^{\theta_i/2}$ for a list of angles $\theta$.
+
+    It uses phase kickback and thus needs a phase gradient state in order to work. This
+    state must be provided externally for efficiency, as it is unaffected and can thus be reused.
+    Refer to [1], section on arbitrary quantum state preparation on page 3.
+
+    Args:
+        n_selections: number of qubits used for encoding the selection register of the QROM, it
```

Comment:
nit: this is called selection_bitsizes elsewhere in the library. It's helpful to keep it consistent.

Reply:
If you're eager you could make this change to other bitsize variables.